### PR TITLE
Optimize OpenApiExtensions.ToConnectorType with an internal cache dur…

### PR DESCRIPTION
**Pull request by LucGenetier**
…ing function initialization

### CODE REVIEW

File: src/PowerFx.ruleset [hunk 1]
```diff
@@ -130,7 +130,7 @@
     <Rule Id="SA1404" Action="Error" />
     <Rule Id="SA1405" Action="Error" />
     <Rule Id="SA1406" Action="Error" />
-    <Rule Id="SA1407" Action="Error" />
+    <Rule Id="SA1407" Action="None" />     <!-- Arithmetic precedence -->
```
Comment (MikeStall) [inline src/PowerFx.ruleset:133]: >None [](http://example.com/codeflow?start=30&length=4)

reminder to put in separate PR

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 2]
```diff
@@ -337,10 +337,35 @@ internal static void WhenPresent(this OpenApiObject apiObj, string str, Action<O
                 action(openApiObj);
             }
         }
+        
+        private static int OpenApiParameterHash(this OpenApiParameter openApiParameter)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:341]: >int [](http://example.com/codeflow?start=23&length=3)

From a _correctness_ point of view, would everything still work if this always returned 0?

File: src/PowerFx.ruleset (hunk 1 above)
Comment (LucGenetier) [reply src/PowerFx.ruleset:133]: Created 1996

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs (hunk 2 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:341]: Absolutely not, it's a hash
But if it would use a static i++, that would work, yes

**Review by MikeStall — CHANGES_REQUESTED** [review]
:clock1:

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 3]
```diff
@@ -337,10 +337,35 @@ internal static void WhenPresent(this OpenApiObject apiObj, string str, Action<O
                 action(openApiObj);
             }
         }
+        
+        private static int OpenApiParameterHash(this OpenApiParameter openApiParameter)
+        {
+            unchecked
+            {
+                // Those are the only parameters that we use from openApiParameter and more specificallyy when traversing objects
+                int hash = 17;
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:346]: Agreed

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiExtensions.GetReturnType(Microsoft.OpenApi.Models.OpenApiOperation)

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 4]
```diff
@@ -574,19 +601,19 @@ public static HttpMethod ToHttpMethod(this OperationType key)
             };
         }
 
-        public static FormulaType GetReturnType(this OpenApiOperation openApiOperation)
-        {
-            (ConnectorType connectorType, string unsupportedReason) = openApiOperation.GetConnectorReturnType();
-            FormulaType ft = connectorType?.FormulaType ?? new BlankType();
-            return ft;
-        }
+        //public static FormulaType GetReturnType(this OpenApiOperation openApiOperation, Dictionary<int, ConnectorType> openApiParameterCache)
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:604]: ah yes!

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiExtensions.GetReturnType(Microsoft.OpenApi.Models.OpenApiOperation)

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs [hunk 5]
```diff
@@ -1111,6 +1106,37 @@ private ConnectorParameterInternals Initialize()
...
+
+                if (x == null || y == null)
+                {
+                    return false;
+                }
+
+                return (x.Schema == y.Schema) && 
+                       (x.Name == y.Name) &&
+                       (x.Required == y.Required) &&
+                       (x.Extensions == y.Extensions);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs:1126]: > (x.Extensions == y.Extensions);        [](http://example.com/codeflow?start=22&length=39)

Extensions is a IDictionary<string, IOpenApiExtension>. 

What does this mean?
